b2luigi/core/utils.py

```diff
@@ -139,11 +139,15 @@ def flatten_to_list_of_dicts(inputs):
     return dict(joined_dict)
 
 
-def task_iterator(task, only_non_complete=False):
+def task_iterator(task, only_non_complete=False, already_seen_tasks = set()):
```

Comment:
However, from a design point of view, I would really consider whether we need/want to extend the interface of the `task_iterator`.  If we never change the `already_seen_tasks` parameter and just always use the default, I would create the set in the function body.

I'm worrying, are there cases where one would expect that the length of the task iterator should be the same as the length of the task list that goes in? Is it unexpected behaviour that the task_iterator also does deduplication? Maybe the deduplication should somehow be separated because it's conceptually separate from iteration? This is something I need to think about or ask for a second opinioin.

Reply:
Thanks for the warning about mutable defaults. This is something that always slips my mind. I'll replace it with a `None`.  
From a quick search of the repo, the `task_iterator` is essentially only used in the batch runner and for a dry run. I think both those use cases will be sped up by the current change. I'm not sure how one would separate the deduplication (besides perhaps making this optional), the main benefit comes in skipping re-traversal of a branch of the graph which requires the deduplication to occur in situ.